theseus/src/launcher/download.rs

```diff
@@ -1,362 +1,222 @@
+        )))?;
+    let path = dirs.version_dir(version).join(format!("{version}.jar"));
+
+    fetch(&client_download.url, Some(&client_download.sha1))
+        .and_then(|it| async move {
+            write(&path, &it).await?;
+            Ok(())
+        })
+        .await
 }
 
 pub async fn download_assets_index(
-    assets_path: &Path,
-    version: &VersionInfo,
-) -> Result<AssetsIndex, LauncherError> {
-    let path =
-        assets_path.join(format!("indexes/{}.json", &version.asset_index.id));
+    dirs: &Dirs,
+    version: &GameVersionInfo,
+) -> crate::Result<AssetsIndex> {
+    let path = dirs
+        .assets_index_dir()
+        .join(format!("{}.json", &version.asset_index.id));
 
     if path.exists() {
-        let content = std::fs::read_to_string(path)?;
-        Ok(serde_json::from_str(&content)?)
+        fs::read(path)
+            .err_into::<crate::Error>()
+            .await
+            .and_then(|ref it| Ok(serde_json::from_slice(it)?))
     } else {
-        let index = fetch_assets_index(version).await?;
-
-        save_file(&path, &bytes::Bytes::from(serde_json::to_string(&index)?))
-            .await?;
-
+        let index = d::minecraft::fetch_assets_index(version).await?;
+        write(&path, &serde_json::to_vec(&index)?).await?;
         Ok(index)
     }
 }
 
 pub async fn download_assets(
-    assets_path: &Path,
-    legacy_path: Option<&Path>,
+    dirs: &Dirs,
+    with_legacy: bool,
     index: &AssetsIndex,
-) -> Result<(), LauncherError> {
-    future::join_all(index.objects.iter().map(|(name, asset)| {
-        download_asset(assets_path, legacy_path, name, asset)
-    }))
-    .await
-    .into_iter()
-    .collect::<Result<Vec<()>, LauncherError>>()?;
-
-    Ok(())
-}
-
-async fn download_asset(
-    assets_path: &Path,
-    legacy_path: Option<&Path>,
-    name: &str,
-    asset: &Asset,
-) -> Result<(), LauncherError> {
-    let hash = &asset.hash;
-    let sub_hash = &hash[..2];
-
-    let mut resource_path = assets_path.join("objects");
-    resource_path.push(sub_hash);
-    resource_path.push(hash);
-
-    let url =
-        format!("https://resources.download.minecraft.net/{sub_hash}/{hash}");
-
-    let resource =
-        save_and_download_file(&resource_path, &url, Some(hash)).await?;
-
-    if let Some(legacy_path) = legacy_path {
-        let resource_path = legacy_path
-            .join(name.replace('/', &std::path::MAIN_SEPARATOR.to_string()));
-        save_file(resource_path.as_path(), &resource).await?;
-    }
+) -> crate::Result<()> {
+    stream::iter(index.objects.iter())
+        .map(Ok::<(&String, &Asset), crate::Error>)
+        .try_for_each_concurrent(None, |(name, asset)| async move {
+            let ref hash = asset.hash;
+            let resource_path = dirs.object_dir(hash);
+            let url = format!(
+                "https://resources.download.minecraft.net/{sub_hash}/hash",
+                sub_hash = &hash[..2]
```

Comment:
This won't work right? `hash` should be `{hash}`

Reply:
Mentioned this in the meeting, it’s fixed locally and was a stupid mistake.